mlir/lib/Conversion/gpu_to_gpu_runtime.cpp

```diff
@@ -0,0 +1,1227 @@
+      auto alloc = mlir::cast<mlir::memref::AllocOp>(it.first);
+      it.second = getAccessType(alloc);
+    }
+
+    auto &block = funcBody.front();
+    for (auto &it : gpuBufferParams) {
+      auto param = block.getArgument(it.first);
+      it.second = getAccessType(param);
+
+      it.second.hostRead = true;
+      it.second.hostWrite = true;
+    }
+
+    mlir::OpBuilder builder(func);
+    for (auto it : gpuBufferAllocs) {
+      auto alloc = mlir::cast<mlir::memref::AllocOp>(it.first);
+      auto access = it.second;
+      auto loc = alloc.getLoc();
+      builder.setInsertionPoint(alloc);
+      auto gpuAlloc = builder.create<mlir::gpu::AllocOp>(
+          loc, alloc.getType(), /*asyncToken*/ nullptr,
+          /*asyncDependencies*/ llvm::None, alloc.dynamicSizes(),
+          alloc.symbolOperands());
+      alloc->replaceAllUsesWith(gpuAlloc);
+      alloc.erase();
+      if (access.hostRead || access.hostWrite)
+        gpuAlloc->setAttr(kGpuAllocShared, builder.getUnitAttr());
+    }
+
+    auto term = block.getTerminator();
+    assert(term);
+
+    llvm::SmallVector<mlir::Value> dims;
+    llvm::SmallPtrSet<mlir::Operation *, 8> filter;
+    for (auto it : gpuBufferParams) {
+      auto param = block.getArgument(it.first);
+      auto access = it.second;
+      auto loc = param.getLoc();
+      builder.setInsertionPointToStart(&block);
+      auto memrefType = param.getType().cast<mlir::MemRefType>();
+      auto rank = static_cast<unsigned>(memrefType.getRank());
+      // dims.resize(rank);
+      filter.clear();
+      for (auto i : llvm::seq(0u, rank)) {
+        if (memrefType.isDynamicDim(i)) {
+          auto op = builder.create<mlir::memref::DimOp>(loc, param, i);
+          dims.push_back(op);
+          filter.insert(op);
+        }
+      }
+      auto allocType = mlir::MemRefType::get(
+          memrefType.getShape(), memrefType.getElementType(),
+          mlir::MemRefLayoutAttrInterface{}, memrefType.getMemorySpace());
+      auto gpuAlloc = builder.create<mlir::gpu::AllocOp>(
+          loc, allocType, /*asyncToken*/ nullptr,
+          /*asyncDependencies*/ llvm::None, dims,
+          /*symbolOperands*/ llvm::None);
+      auto allocResult = gpuAlloc.getResult(0);
+
+      if (access.hostRead || access.hostWrite)
+        gpuAlloc->setAttr(kGpuAllocShared, builder.getUnitAttr());
+
+      if (access.hostWrite && access.deviceRead) {
+        auto copy =
+            builder.create<mlir::memref::CopyOp>(loc, param, allocResult);
+        filter.insert(copy);
+      }
+
+      if (allocType != memrefType) {
+        allocResult =
+            builder.create<mlir::memref::CastOp>(loc, allocResult, memrefType);
+      }
+
+      param.replaceAllUsesExcept(allocResult, filter);
+      builder.setInsertionPoint(term);
+      if (access.hostRead && access.deviceWrite)
+        builder.create<mlir::memref::CopyOp>(loc, allocResult, param);
+
+      // builder.create<mlir::memref::DeallocOp>(loc, allocResult);
```

Comment:
We will need to add an option to pass to generate either `gpu.dealloc` or `memref.dealloc`. For now can you add option to generate either `memref.dealloc` or nothing.